### ISSUE TITLE
Research extraction alternatives: local NER/RE models vs LLM extraction

### ISSUE THREAD

**Issue by zonk1024** [ourochronos/covalence#42]
## Context

Current extraction uses LLM chat completions (OpenAI-compatible) for entity and relationship extraction. This works but is expensive at scale. Moving toward local SLMs for the entire ingestion pipeline.

## Current State
- **GPT-4o via OpenRouter**: Works with `response_format: json_object`. $2.50/M input.
- **Gemini 2.5 Flash via OpenRouter**: Confirmed working. ~$0.30/M input.
- **GLiNER2**: Already stubbed in `gliner_extractor.rs`. Zero-shot NER, runs on CPU.
- **Two-pass extraction**: Already designed in `two_pass_extractor.rs` (GLiNER → LLM for edges).
- **Coreference**: Module stubbed in `coreference.rs` but not wired.

## Target Pipeline

```
ReaderLM-v2 (convert) → Fastcoref (coref) → GLiNER2 (entities) → NuExtract/Flash (edges) → HDBSCAN (normalize)
```

### Entity Extraction (NER)
| Model | Size | Type | Notes |
|-------|------|------|-------|
| **GLiNER2** | ~500MB | Zero-shot NER | Already stubbed. Best-in-class for zero-shot. |
| Fine-tuned BERT NER | ~500MB | Supervised | Domain-specific, requires training data |

### Coreference Resolution (Pre-extraction)
| Model | Size | Type | Notes |
|-------|------|------|-------|
| **Fastcoref** (RoBERTa) | ~500MB | Coref | Resolves pronouns before NER. Huge quality win. |
| SpanBERT | ~500MB | Coref | Alternative |

### Relationship Extraction
| Model | Size | Type | Notes |
|-------|------|------|-------|
| **NuExtract 2.0** (NuMind) | 2-4B | Structured JSON extraction | Purpose-built for extraction, beats larger models |
| Qwen-2.5-3B-Instruct | 3B | General + grammar constraints | With Outlines/llama.cpp JSON grammar |
| Gemini 2.5 Flash | Cloud | Via OpenRouter | Confirmed working, $0.30/M, good fallback |
| REBEL (BART-based) | ~600MB | End-to-end RE | Own ontology, may conflict with open-ended edges |

### Embeddings (Local)
| Model | Size | Dims | Context | Notes |
|-------|------|------|---------|-------|
| **Nomic-Embed-Text 1.5** | ~550MB | 768 | 8192 | Top-tier open-weight, quantizable |
| BGE-M3 (BAAI) | ~1GB | 1024 | 8192 | Multi-lingual, hybrid retrieval |
| Jina Embeddings v3 | ~1GB | 1024 | 8192 | Also available as API |

### Content Conversion
| Model | Size | Notes |
|-------|------|-------|
| **ReaderLM-v2** | 1.5B | ✅ Already running on MLX (#41) |

### Mac Mini (M4, 16GB) — always-on server
- Covalence + Postgres + petgraph
- GLiNER2 (~1GB), Fastcoref (~1GB), Nomic-Embed (~1GB) — small models, lazy-loaded
- ReaderLM-v2 (~3GB) — already running

### Gaming Desktop (R9 7945HX, 96GB DDR5, RX 9070 16GB) — burst inference
- NuExtract 2B (~4GB), larger models
- NOT always-on (shared with gaming/dev use)
- Could serve as inference sidecar over LAN when available
- ROCm for GPU acceleration

## Architecture Pattern
Same HTTP sidecar pattern as #41 (ReaderLM). Each model behind a FastAPI endpoint. Covalence calls `COVALENCE_*_URL` env vars. If sidecar is down, fall back to cloud API or skip gracefully.

## Priority Order
1. **GLiNER2** — already stubbed, wire it up for entity extraction
2. **Fastcoref** — coref resolution before extraction (new sidecar)
3. **NuExtract or Gemini Flash** — relationship extraction (replace GPT-4o)
4. **Nomic-Embed** — local embeddings (eliminate API cost)

## Research Questions
- Does NuExtract handle open-ended edge types well, or does it need enum constraints?
- How much does Fastcoref improve downstream entity/edge quality? (A/B test)
- Can Nomic-Embed match Voyage/OpenAI quality on our technical corpus?
- ROCm compatibility with 9070 for NuExtract/larger models?

## Related
- #5 — GLiNER2 integration
- #11 — Fine-tune relationship extraction
- #32 — Two-pass extraction
- #36 — HDBSCAN ontology clustering
- #37 — Cluster pinning (gravity wells)
- #41 — ReaderLM-v2 sidecar (pattern to follow)
- `design/pipeline-tuning.md` — Full pipeline documentation

**Comment by zonk1024** [comment]
## Local Model Testing Progress

### ✅ ReaderLM-v2 (HTML→Markdown conversion)
- **Model**: `mlx-community/jinaai-ReaderLM-v2` (1.5B params)
- **Runtime**: MLX on Apple Silicon
- **Venv**: `.venv/` (Python 3.14)
- **Sidecar**: `sidecar/readerlm/server.py` on port 8432
- **Result**: Excellent markdown output including tables, headings, lists
- **Status**: Wired into Covalence via #41 ✅

### ✅ Fastcoref (Coreference Resolution)
- **Model**: `biu-nlp/f-coref` (90M params, RoBERTa-based)
- **Runtime**: PyTorch CPU (MPS had issues)
- **Venv**: `.venv-312/` (Python 3.12 — spaCy incompatible with 3.14)
- **Speed**: 100+ texts/sec on CPU
- **Dependency note**: Requires `transformers<5` (conflicts with mlx-lm, hence separate venv)
- **Result**: Perfect pronoun resolution:
  ```
  ORIGINAL: Audun Jøsang developed Subjective Logic. He published it in his 2016 book.
  CLUSTERS: [[Audun Jøsang, He, his], [Subjective Logic, it, It]]
  
  ORIGINAL: Tim Cook took the stage at WWDC. He announced the new M4 chip. It delivers unprecedented performance.
  CLUSTERS: [[Tim Cook, He], [the new M4 chip, It, its]]
  ```
- **Status**: Tested, needs sidecar + Covalence wiring

### 🔜 GLiNER2 (NER) — next up
### 🔜 NuExtract (Relationship Extraction) — after GLiNER2
### 🔜 Nomic-Embed (Local Embeddings) — after NuExtract

**Comment by zonk1024** [comment]
### ✅ GLiNER2 (Named Entity Recognition)
- **Model**: `urchade/gliner_medium-v2.1` (~500MB, ONNX)
- **Runtime**: CPU (ONNX Runtime)
- **Venv**: `.venv-312/` (Python 3.12)
- **Labels tested**: person, organization, technology, concept, event, location, algorithm
- **Threshold**: 0.4
- **Result**: Excellent accuracy across all test texts:
  ```
  TEXT: Audun Jøsang developed Subjective Logic...
    person    0.99  Audun Jøsang
    concept   0.81  Subjective Logic
    concept   0.71  Dempster-Shafer theory
    concept   0.67  epistemic opinions
    concept   0.73  Dirichlet distributions

  TEXT: The HNSW algorithm, published by Malkov and Yashunin...
    algorithm 0.87  HNSW algorithm
    person    0.93  Malkov
    person    0.91  Yashunin
    technology 0.70  pgvector extension

  TEXT: HDBSCAN clustering, developed by Campello, Moulavi, and Sander...
    algorithm 0.81  HDBSCAN clustering
    person    0.98  Campello
    person    0.97  Moulavi
    person    0.97  Sander
    organization 0.67  Covalence
  ```
- **Notes**: Catches more fine-grained concepts than GPT-4o (e.g. "epistemic opinions", "Dirichlet distributions"). Zero-shot, no training needed. Already stubbed in `gliner_extractor.rs`.
- **Status**: Tested, needs sidecar wiring

### 🔜 NuExtract (Relationship Extraction) — testing next

**Comment by zonk1024** [comment]
### ✅ NuExtract-1.5-tiny (Relationship Extraction)
- **Model**: `numind/NuExtract-1.5-tiny` (0.5B, Qwen2-based)
- **Runtime**: PyTorch CPU (float32)
- **Venv**: `.venv-312/` (Python 3.12)
- **Schema**: Open-ended — no enum constraints on relationship types
- **Result**: Perfect structured JSON extraction:
  ```json
  {
    "relationships": [
      {"source_entity": "Audun Jøsang", "target_entity": "Subjective Logic",
       "relationship_type": "developed", "description": "a mathematical framework..."},
      {"source_entity": "Subjective Logic", "target_entity": "Dempster-Shafer theory",
       "relationship_type": "extends", "description": "with epistemic opinions..."},
      {"source_entity": "Subjective Logic", "target_entity": "property graph",
       "relationship_type": "annotate", "description": "with epistemic confidence scores"}
    ]
  }
  ```
- **Notes**: Handles open-ended edge types perfectly — no enum constraints needed. HDBSCAN will normalize the edge type vocabulary post-extraction. 0.5B fits anywhere.
- **Status**: Tested, needs sidecar wiring

---

## 🎉 FULLY LOCAL PIPELINE CONFIRMED

| Stage | Model | Params | RAM | Status |
|-------|-------|--------|-----|--------|
| Convert | ReaderLM-v2 (MLX) | 1.5B | ~3GB | ✅ Running as sidecar |
| Coref | Fastcoref (f-coref) | 90M | ~500MB | ✅ Tested |
| NER | GLiNER2 (ONNX) | ~500MB | ~1GB | ✅ Tested |
| Relations | NuExtract-1.5-tiny | 0.5B | ~1GB | ✅ Tested |
| **Total** | | **~2.6B** | **~5.5GB** | **Fits on Mac Mini (16GB)** |

**Zero API costs for entire extraction pipeline.** Gemini Flash becomes optional quality upgrade, not a dependency.

### Next: Sidecar architecture to wire all four models into Covalence

**Comment by zonk1024** [comment]
Closed. We have definitively chosen Gemini Flash 3.0 via a two-pass extraction pipeline. Local models lack the context window required for statement-first extraction.
